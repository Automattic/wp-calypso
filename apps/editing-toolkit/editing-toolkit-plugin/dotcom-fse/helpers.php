<?php
/**
 * Helpers for Full Site Editing.
 *
 * This file is always loaded, so these functions should always exist if the
 * plugin is activated on the site. (Not to be confused with whether FSE is
 * active on the site!)
 *
 * @package A8C\FSE
 */

namespace A8C\FSE;

/**
 * NOTE: In most cases, you should NOT use this function. Please use
 * load_full_site_editing instead. This function should only be used if you need
 * to include the FSE files somewhere like a script. I.e. if you want to access
 * a class defined here without needing full FSE functionality.
 */
function dangerously_load_full_site_editing_files() {
	require_once __DIR__ . '/blocks/navigation-menu/index.php';
	require_once __DIR__ . '/blocks/post-content/index.php';
	require_once __DIR__ . '/blocks/site-description/index.php';
	require_once __DIR__ . '/blocks/site-title/index.php';
	require_once __DIR__ . '/blocks/template/index.php';
	require_once __DIR__ . '/class-full-site-editing.php';
	require_once __DIR__ . '/templates/class-rest-templates-controller.php';
	require_once __DIR__ . '/templates/class-wp-template.php';
	require_once __DIR__ . '/templates/class-wp-template-inserter.php';
	require_once __DIR__ . '/templates/class-template-image-inserter.php';
	require_once __DIR__ . '/serialize-block-fallback.php';
}

/**
 * Whether or not FSE is active.
 * If false, FSE functionality should be disabled.
 *
 * @returns bool True if FSE is active, false otherwise.
 */
function is_full_site_editing_active() {
	/**
	 * There are times when this function is called from the WordPress.com public
	 * API context. In this case, we need to switch to the correct blog so that
	 * the functions reference the correct blog context.
	 */
	$multisite_id  = apply_filters( 'a8c_fse_get_multisite_id', false );
	$should_switch = is_multisite() && $multisite_id;
	if ( $should_switch ) {
		switch_to_blog( $multisite_id );
	}

	$is_active = is_site_eligible_for_full_site_editing() && is_theme_supported() && did_insert_template_parts();

	if ( $should_switch ) {
		restore_current_blog();
	}
	return $is_active;
}

/**
 * Returns the slug for the current theme.
 *
 * This even works for the WordPress.com API context where the current theme is
 * not correct. The filter correctly switches to the correct blog context if
 * that is the case.
 *
 * @return string Theme slug.
 */
function get_theme_slug() {
	/**
	 * Used to get the correct theme in certain contexts.
	 *
	 * For example, in the wpcom API context, the theme slug is a8c/public-api,
	 * so we need to grab the correct one with the filter.
	 *
	 * @since 0.7
	 *
	 * @param string current theme slug is the default if nothing overrides it.
	 */
	return apply_filters( 'a8c_fse_get_theme_slug', get_stylesheet() );
}

/**
 * Returns a normalized slug for the current theme.
 *
 * In some cases, the theme is located in a subfolder like `pub/maywood`. Use
 * this function to get the slug without the prefix.
 *
 * @param string $theme_slug The raw theme_slug to normalize.
 * @return string Theme slug.
 */
function normalize_theme_slug( $theme_slug ) {
	// Normalize the theme slug.
	if ( 'pub/' === substr( $theme_slug, 0, 4 ) ) {
		$theme_slug = substr( $theme_slug, 4 );
	}

	if ( '-wpcom' === substr( $theme_slug, -6, 6 ) ) {
		$theme_slug = substr( $theme_slug, 0, -6 );
	}

	return $theme_slug;
}

/**
 * Whether or not the site is eligible for FSE. This is essentially a feature
 * gate to disable FSE on some sites which could theoretically otherwise use it.
 *
 * By default, sites should not be eligible.
 *
 * @return bool True if current site is eligible for FSE, false otherwise.
 */
function is_site_eligible_for_full_site_editing() {
	/**
	 * Can be used to disable Full Site Editing functionality.
	 *
	 * @since 0.2
	 *
	 * @param bool true if Full Site Editing should be disabled, false otherwise.
	 */
	return ! apply_filters( 'a8c_disable_full_site_editing', true );
}

/**
 * Whether or not current theme is enabled for FSE.
 *
 * @return bool True if current theme supports FSE, false otherwise.
 */
function is_theme_supported() {
	if ( is_multisite() && 0 === get_current_blog_id() ) {
		// get_theme_slug will always return false.
		return false;
	}
	// Use un-normalized theme slug because get_theme requires the full string.
	$theme = wp_get_theme( get_theme_slug() );
	return ! $theme->errors() && in_array( 'full-site-editing', $theme->tags, true );
}

/**
 * Determines if the template parts have been inserted for the current theme.
 *
 * We want to gate on this check in is_full_site_editing_active so that we don't
 * load FSE for sites which did not get template parts for some reason or another.
 *
 * For example, if a user activates theme A on their site and gets FSE, but then
 * activates theme B which does not have FSE, they will not get FSE flows. If we
 * retroactively add FSE support to theme B, the user should not get FSE flows
 * because their site would be modified. Instead, FSE flows would become active
 * when they specifically take action to re-activate the theme.
 *
 * @return bool True if the template parts have been inserted. False otherwise.
 */
function did_insert_template_parts() {
	require_once __DIR__ . '/templates/class-wp-template-inserter.php';

	$theme_slug = normalize_theme_slug( get_theme_slug() );
	$inserter   = new WP_Template_Inserter( $theme_slug );
	return $inserter->is_template_data_inserted();
}

/**
 * Inserts default full site editing data for current theme on plugin/theme activation.
 *
 * We put this here outside of the normal FSE class because FSE is not active
 * until the template parts are inserted. This makes sure we insert the template
 * parts when switching to a theme which supports FSE.
 *
 * This will populate the default header and footer for current theme, and create
 * About and Contact pages. Nothing will populate if the data already exists, or
 * if the theme is unsupported.
 */
function populate_wp_template_data() {
	if ( ! is_theme_supported() ) {
		return;
	}

	require_once __DIR__ . '/templates/class-template-image-inserter.php';
	require_once __DIR__ . '/templates/class-wp-template-inserter.php';

	$theme_slug = normalize_theme_slug( get_theme_slug() );

	$template_inserter = new WP_Template_Inserter( $theme_slug );
	$template_inserter->insert_default_template_data();
	$template_inserter->insert_default_pages();
}
register_activation_hook( __FILE__, __NAMESPACE__ . '\populate_wp_template_data' );
add_action( 'switch_theme', __NAMESPACE__ . '\populate_wp_template_data' );
