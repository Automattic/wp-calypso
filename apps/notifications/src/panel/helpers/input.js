export const keys = {
	KEY_A: 65,
	KEY_E: 69,
	KEY_L: 76,
	KEY_S: 83,
	KEY_T: 84,
};

export const modifierKeyIsActive = ( e ) => {
	return e.altKey || e.ctrlKey || e.metaKey;
};
