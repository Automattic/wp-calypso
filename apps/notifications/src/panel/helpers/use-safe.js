/**
 * External dependencies
 */
import { useCallback, useRef, useLayoutEffect } from 'react';

/**
 * Hook to be used to make sure a function `fn` is called only
 * if the component which uses it is still mounted.
 *
 * @param {Function} fn A function you want to be safe to call
 */
export default function useSafe( fn ) {
	const mounted = useRef( false );

	useLayoutEffect( () => {
		mounted.current = true;
		return () => ( mounted.current = false );
	}, [] );

	return useCallback( ( ...args ) => ( mounted.current ? fn( ...args ) : void 0 ), [ fn ] );
}
