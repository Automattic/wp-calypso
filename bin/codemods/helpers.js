function bindEvents( jscodeshiftProcess ) {
	jscodeshiftProcess.stdout.on( 'data', ( data ) => {
		process.stdout.write( data );
	} );

	jscodeshiftProcess.stderr.on( 'data', ( data ) => {
		process.stderr.write( data );
	} );
}

module.exports = {
	bindEvents,
};
