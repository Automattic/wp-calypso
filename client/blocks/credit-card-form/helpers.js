/**
 * External dependencies
 */
import { useState, useEffect, useRef } from 'react';
import { camelCase, kebabCase, debounce } from 'lodash';

/**
 * Internal dependencies
 */
import notices from 'notices';
import { handleRenewNowClick, isRenewable } from 'lib/purchases';
import wpcomFactory from 'lib/wp';

const wpcom = wpcomFactory.undocumented();

export async function saveCreditCard( {
	createCardToken,
	saveStoredCard,
	translate,
	successCallback,
	apiParams,
	purchase,
	siteSlug,
	formFieldValues,
} ) {
	const cardDetails = kebabCaseFormFields( formFieldValues );
	const { token } = await createCardTokenAsync( { cardDetails, createCardToken } );

	if ( saveStoredCard ) {
		await saveStoredCard( { token } );
		notices.success( translate( 'Card added successfully' ), {
			persistent: true,
		} );
		successCallback();
		return;
	}

	const updatedCreditCardApiParams = getParamsForApi( cardDetails, token, apiParams );
	const response = wpcom.updateCreditCard( updatedCreditCardApiParams );
	if ( response.error ) {
		throw new Error( response );
	}

	if ( purchase && siteSlug && isRenewable( purchase ) ) {
		const noticeMessage = translate(
			'Your credit card details were successfully updated, but your subscription has not been renewed yet.'
		);
		const noticeOptions = {
			button: translate( 'Renew Now' ),
			onClick: function( event, closeFunction ) {
				handleRenewNowClick( purchase, siteSlug );
				closeFunction();
			},
			persistent: true,
		};
		notices.info( noticeMessage, noticeOptions );
		successCallback();
		return;
	}
	notices.success( response.success, {
		persistent: true,
	} );
	successCallback();
}

export function getParamsForApi( cardDetails, cardToken, extraParams = {} ) {
	return {
		...extraParams,
		country: cardDetails.country,
		zip: cardDetails[ 'postal-code' ],
		month: cardDetails[ 'expiration-date' ].split( '/' )[ 0 ],
		year: cardDetails[ 'expiration-date' ].split( '/' )[ 1 ],
		name: cardDetails.name,
		document: cardDetails.document,
		street_number: cardDetails[ 'street-number' ],
		address_1: cardDetails[ 'address-1' ],
		address_2: cardDetails[ 'address-2' ],
		city: cardDetails.city,
		state: cardDetails.state,
		phone_number: cardDetails[ 'phone-number' ],
		cardToken,
	};
}

export function useDebounce( value, delay ) {
	const [ debouncedValue, setDebouncedValue ] = useState( value );
	const debounced = useRef( debounce( newValue => setDebouncedValue( newValue ), delay ) );
	useEffect( () => debounced.current( value ), [ value ] );
	return [ debouncedValue, setDebouncedValue ];
}

function createCardTokenAsync( { cardDetails, createCardToken } ) {
	return new Promise( ( resolve, reject ) => {
		createCardToken( cardDetails, ( gatewayError, gatewayData ) => {
			if ( gatewayError || ! gatewayData.token ) {
				reject( gatewayError || new Error( 'No card token returned' ) );
				return;
			}
			resolve( gatewayData );
		} );
	} );
}

export function areFormFieldsEmpty( formFieldValues ) {
	return Object.keys( formFieldValues ).reduce( ( isEmpty, key ) => {
		return formFieldValues[ key ].length ? false : isEmpty;
	}, true );
}

export function kebabCaseFormFields( formFieldValues ) {
	return Object.keys( formFieldValues ).reduce( ( fields, key ) => {
		fields[ kebabCase( key ) ] = formFieldValues[ key ];
		return fields;
	}, {} );
}

export function camelCaseFormFields( formFieldValues ) {
	return Object.keys( formFieldValues ).reduce( ( fields, key ) => {
		fields[ camelCase( key ) ] = formFieldValues[ key ];
		return fields;
	}, {} );
}

export function assignAllFormFields( formFieldValues, value ) {
	return Object.keys( formFieldValues ).reduce( ( fields, key ) => {
		fields[ key ] = value;
		return fields;
	}, {} );
}

export function getInitializedFields( initialValues = {} ) {
	const fieldNames = [
		'name',
		'number',
		'cvv',
		'expirationDate',
		'country',
		'postalCode',
		'streetNumber',
		'address1',
		'address2',
		'phoneNumber',
		'streetNumber',
		'city',
		'state',
		'document',
		'brand',
	];
	return fieldNames.reduce( ( finalFields, fieldName ) => {
		return { ...finalFields, ...{ [ fieldName ]: initialValues[ fieldName ] || '' } };
	}, {} );
}
