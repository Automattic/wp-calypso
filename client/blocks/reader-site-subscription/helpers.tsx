import { Reader } from '@automattic/data-stores';
import { getCurrencyObject } from '@automattic/format-currency';
import moment from 'moment';

export type SiteSubscriptionDetailsProps = {
	subscriptionId: number;
	subscriberCount: number;
	dateSubscribed: string;
	siteIcon: string | null;
	name: string;
	blogId: number;
	feedId: number;
	deliveryMethods: Reader.SiteSubscriptionDeliveryMethods;
	url: string;
	paymentDetails: Reader.SiteSubscriptionPaymentDetails[];
};

export type PaymentPlan = {
	id: string;
	renewalPrice: string;
	renewalDate: string;
};

export const getPaymentInterval = ( renew_interval: string ) => {
	if ( renew_interval === null ) {
		return 'one time';
	} else if ( renew_interval === '1 month' ) {
		return 'per month';
	} else if ( renew_interval === '1 year' ) {
		return 'per year';
	}
};

export function formatRenewalPrice( renewalPrice: string, currency: string ) {
	if ( ! renewalPrice ) {
		return '';
	}

	const money = getCurrencyObject( parseFloat( renewalPrice ), currency );
	return money.integer !== '0' ? `${ money.symbol }${ money.integer } /` : '';
}

export function formatRenewalDate( renewalDate: string, localeSlug: string ) {
	if ( ! renewalDate ) {
		return '';
	}

	const date = moment( renewalDate );
	return date.locale( localeSlug ).format( 'LL' );
}
