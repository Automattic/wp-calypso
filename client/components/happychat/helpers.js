/**
 * External dependencies
 */
import React from 'react';

/**
 * Internal dependencies
 */
import {
	propExists,
	propEquals,
	when
} from './functional';
import Timeline from './timeline';
import Composer from './composer';
import Spinner from 'components/spinner';

/*
 * Renders a spinner in a flex-box context so it is centered vertically and horizontally
 */
const renderLoading = () => (
	<div className="happychat__loading">
		<Spinner />
	</div>
);

export const isAvailable = propExists( 'available' );
export const isConnecting = propEquals( 'connectionStatus', 'connecting' );
export const isConnected = propEquals( 'connectionStatus', 'connected' );

export const timeline = when(
	isConnecting,
	renderLoading,
	( { onScrollContainer } ) => <Timeline onScrollContainer={ onScrollContainer } />
);

export const composer = when( isConnected, () => <Composer /> );
