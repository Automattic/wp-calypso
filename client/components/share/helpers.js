/**
 * External dependencies
 */
import { find } from 'lodash';

//Mostly copied from Seo Preview

export const shortEnough = limit => title => title.length <= limit ? title : false;

export const truncatedAtSpace = ( lower, upper ) => fullTitle => {
	const title = fullTitle.slice( 0, upper );
	const lastSpace = title.lastIndexOf( ' ' );

	return ( lastSpace > lower && lastSpace < upper )
		? title.slice( 0, lastSpace ).concat( '…' )
		: false;
};

export const hardTruncation = limit => title => title.slice( 0, limit ).concat( '…' );

export const firstValid = ( ...predicates ) => a =>
	find( predicates, ( p => false !== p( a ) ) )( a );
