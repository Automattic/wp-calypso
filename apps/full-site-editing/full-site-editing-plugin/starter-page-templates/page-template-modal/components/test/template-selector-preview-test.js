/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import { blocksFixture } from './helpers/templates-blocks-helpers';
import TemplateSelectorPreview from '../template-selector-preview';

// Mock the "pass through" version of the `BlockPreview` component
// See `components/block-preview.js`
jest.mock( '../block-preview', () => () => {
	return <div data-testid="block-template-preview">MockedBlockPreview</div>;
} );

// Required to handle `BlockPreview` usage of Mutation Observer
beforeAll(
	( global.MutationObserver = jest.fn( () => {
		return {
			observe: jest.fn(),
			disconnect: jest.fn(),
		};
	} ) )
);

afterAll( () => {
	global.MutationObserver.mockRestore();
} );

describe( 'TemplateSelectorPreview', () => {
	describe( 'Basic rendering', () => {
		it( 'renders the preview when blocks are provided', () => {
			const { queryByTestId, container } = render(
				<TemplateSelectorPreview blocks={ blocksFixture } viewportWidth={ 960 } />
			);

			expect( queryByTestId( 'block-template-preview' ) ).toBeInTheDocument();
			expect( container ).toMatchSnapshot();
		} );

		it( 'renders placeholder when no blocks are provided', () => {
			const { getByText, queryByTestId } = render(
				<TemplateSelectorPreview viewportWidth={ 960 } />
			);

			expect( getByText( 'Select a page template to preview.' ) ).toBeInTheDocument();
			expect( queryByTestId( 'block-template-preview' ) ).not.toBeInTheDocument();
		} );

		it( 'renders placeholder when blocks is not an array', () => {
			const invalidBlocksProp = {
				'some-block-1': {
					block: 'foo',
				},
				'some-block-2': {
					block: 'bar',
				},
			};
			const { getByText, queryByTestId } = render(
				<TemplateSelectorPreview blocks={ invalidBlocksProp } viewportWidth={ 960 } />
			);

			expect( getByText( 'Select a page template to preview.' ) ).toBeInTheDocument();
			expect( queryByTestId( 'block-template-preview' ) ).not.toBeInTheDocument();
		} );
	} );
} );
