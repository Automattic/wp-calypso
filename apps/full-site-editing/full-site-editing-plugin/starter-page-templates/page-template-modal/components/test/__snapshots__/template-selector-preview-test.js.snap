// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateSelectorPreview Basic rendering renders the preview when blocks are provided 1`] = `
<div>
  <div
    class="template-selector-preview editor-styles-wrapper"
  >
    <div
      class="components-disabled"
    >
      <div
        class="edit-post-visual-editor"
      >
        <div
          class="editor-styles-wrapper"
          style="visibility: hidden;"
        >
          <div
            class="editor-writing-flow"
          >
            <div
              class="editor-post-title"
              style="transform: none;"
            >
              <div
                class="editor-post-title__block"
              >
                <textarea
                  class="editor-post-title__input"
                />
              </div>
            </div>
            <div
              data-testid="block-template-preview"
            >
              MockedBlockPreview
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
